### PR TITLE
[diffusion] [NPU] Optimize LTX-2/2.3 inference performance for NPU

### PR DESCRIPTION
**Pull request by e-martirosian**
## Motivation

NPU compatibility and performance optimizations for LTX-2/2.3 inference, along with fixing crashes on the main branch.

## Modifications

Performance improvements:

- Converted all `float64` computations to `float32` for NPU (NPU does not support float64 precision)
- Use sglang's NPU-optimized `RMSNorm` instead of `torch.nn.RNSNorm`
- Reuse CUDA Triton kernel for NPU paths

## T2V Accuracy Tests

Fails in the main branch on 14 Aug. 
Using 5899674504af7a12109332a0d14a96f1c5d5bb9d for the baseline.

Prompt: `A majestic black horse gallops freely across a vast golden meadow at sunset. Its long mane and tail flow dramatically in the wind as it runs toward the camera, kicking up small clouds of dust. Warm golden sunlight creates beautiful rim lighting around the horse, with mountains and a glowing orange sky in the background. The camera smoothly tracks alongside the horse, then slowly moves into a close-up of its face. Ultra-realistic horse anatomy and natural movement, cinematic lighting, detailed fur, realistic muscles, subtle wind movement, shallow depth of field, epic cinematic atmosphere, smooth camera motion, photorealistic, no text, no humans, no distortion.`

### Before
<details>
<summary>LTX-2</summary>
<h4>LTX2Pipeline</h4>

https://github.com/user-attachments/assets/d0e01b42-997a-4682-b108-78c2b9d62483

<h4>LTX2TwoStagePipeline</h4>

https://github.com/user-attachments/assets/cd72613a-52de-4f97-8862-8b4e78c2ba3d

</details>

<details>
<summary>LTX-2.3</summary>
<h4>LTX2Pipeline</h4>

https://github.com/user-attachments/assets/d42b2023-b9a2-405c-86b5-2069014daee0

<h4>LTX2TwoStagePipeline</h4>

https://github.com/user-attachments/assets/f0c7533c-bc31-44ac-b348-822d95834731

<h4>❌ LTX2TwoStageHQPipeline</h4>

</details>

### After

<details>
<summary>LTX-2</summary>
<h4>LTX2Pipeline</h4>

https://github.com/user-attachments/assets/4c7a24e9-f4e8-412a-aac6-350de459d242

<h4>LTX2TwoStagePipeline</h4>

https://github.com/user-attachments/assets/9eee8e53-b7cb-4d72-be16-6d13ec0c5ea2

</details>
<details>
<summary>LTX-2.3</summary>
<h4>LTX2Pipeline</h4>

https://github.com/user-attachments/assets/38a99cfc-45fa-4a66-9489-3b7f68d119ff

<h4>LTX2TwoStagePipeline</h4>

https://github.com/user-attachments/assets/4a2827f4-6d7d-4f19-9431-9e218ef6ba62

<h4>LTX2TwoStageHQPipeline</h4>

https://github.com/user-attachments/assets/4cb190e2-90cd-4b8f-b26b-3758753a17d9

</details>

## TI2V Accuracy Tests

Fails in the main branch on 14 Aug. 
Using 5899674504af7a12109332a0d14a96f1c5d5bb9d for the baseline.

Prompt: `Animate the horse naturally and realistically while preserving its exact appearance, body proportions, colors, markings, and the original scene. The horse comes alive, gently moves its head, blinks, breathes naturally, flicks its ears, and swishes its tail. It then slowly begins walking forward with realistic horse movement and natural leg coordination. Add subtle mane and hair movement from the wind. Keep the camera smooth and cinematic, with realistic lighting and physically accurate motion. The horse remains the main focus throughout the video. No changes to the horse's identity or appearance, no deformation, no extra limbs, no unnatural movements.`

Image:

hourse

### Before
<details>
<summary>LTX-2</summary>
<h4>LTX2Pipeline</h4>

https://github.com/user-attachments/assets/5591b290-082d-4f83-a0d2-861c1dabc6c0

<h4>LTX2TwoStagePipeline</h4>

https://github.com/user-attachments/assets/ef323fd5-56da-4dfc-a659-abfc5765f27a

</details>

<details>
<summary>LTX-2.3</summary>
<h4>LTX2Pipeline</h4>

https://github.com/user-attachments/assets/6713f155-a27e-482a-9488-8550207032f6

<h4>LTX2TwoStagePipeline</h4>

https://github.com/user-attachments/assets/bd24ef9b-7b77-4c6d-b8d7-2df35bf30bcb

<h4>❌ LTX2TwoStageHQPipeline</h4>

</details>

### After

<details>
<summary>LTX-2</summary>
<h4>LTX2Pipeline</h4>

https://github.com/user-attachments/assets/49300506-61f2-4177-bef9-eb697623c4ff

<h4>LTX2TwoStagePipeline</h4>

https://github.com/user-attachments/assets/c67f02b6-d136-4be2-add4-4dff41cc108f

</details>

<details>
<summary>LTX-2.3</summary>
<h4>LTX2Pipeline</h4>

https://github.com/user-attachments/assets/3f9d668d-04a5-463b-a762-3be79603ede7

<h4>LTX2TwoStagePipeline</h4>

https://github.com/user-attachments/assets/f2fff727-3534-44de-b829-62749a66a7d4

<h4>LTX2TwoStageHQPipeline</h4>

https://github.com/user-attachments/assets/ef5b90f0-ab7a-4607-966e-9ab664fee251

</details>

## Speed Tests and Profiling - A3

Server `LTX2Pipeline|LTX2TwoStagePipeline`:

```
PYTORCH_NPU_ALLOC_CONF=expandable_segments:True \
SGLANG_CACHE_DIT_ENABLED=true SGLANG_CACHE_DIT_FN=2 \
SGLANG_CACHE_DIT_BN=1\SGLANG_CACHE_DIT_WARMUP=4 \
SGLANG_CACHE_DIT_RDT=0.4 SGLANG_CACHE_DIT_MC=4 \
SGLANG_CACHE_DIT_TAYLORSEER=true SGLANG_CACHE_DIT_TS_ORDER=2 \
sglang serve --model-path Lightricks/(LTX-2|LTX-2.3) \
 --pipeline-class-name (LTX2Pipeline|LTX2TwoStagePipeline) \
 --port 30000
```

Server `LTX2TwoStageHQPipeline`:

```
PYTORCH_NPU_ALLOC_CONF=expandable_segments:True \
sglang serve --model-path Lightricks/LTX-2.3 \
 --pipeline-class-name LTX2TwoStageHQPipeline \
 --port 30000 \
 --dit-layerwise-offload
```

T2V client:

```
python3 -m sglang.multimodal_gen.benchmarks.bench_serving --backend sglang-video --dataset vbench \
--task "text-to-video" --num-prompts 20 --max-concurrency 1
```

I2V client:

```
python3 -m sglang.multimodal_gen.benchmarks.bench_serving --backend sglang-video --dataset vbench \
--task "image-to-video" --num-prompts 20 --max-concurrency 1
```

#### LTX-2

<table>
  <tr>
    <th>Pipeline</th>
    <th>Latency Before, ms</th>
    <th>Latency After, ms</th>
    <th>Improvement, %</th>
  </tr>
  <tr>
    <td rowspan="1">LTX2Pipeline</td>
    <td>33.08</td>
    <td>30.07</td>
    <td>🟢 9.1</td>
  </tr>
  <tr>
    <td rowspan="1">LTX2TwoStagePipeline</td>
    <td>16.39</td>
    <td>15.14</td>
    <td>🟢 7.6</td>
  </tr>
</table>

#### LTX-2.3

<table>
  <tr>
    <th>Pipeline</th>
    <th>Latency Before, ms</th>
    <th>Latency After, ms</th>
    <th>Improvement, %</th>
  </tr>
  <tr>
    <td rowspan="1">LTX2Pipeline</td>
    <td>51.10</td>
    <td>47.11</td>
    <td>🟢 7.8</td>
  </tr>
  <tr>
    <td rowspan="1">LTX2TwoStagePipeline</td>
    <td>23.06</td>
    <td>20.55</td>
    <td>🟢 10.9</td>
  </tr>
  <tr>
    <td rowspan="1">LTX2TwoStageHQPipeline</td>
    <td>❌</td>
    <td>230.04</td>
    <td>⚪ N/A</td>
  </tr>
</table>

#### LTX-2

<table>
  <tr>
    <th>Pipeline</th>
    <th>Latency Before, ms</th>
    <th>Latency After, ms</th>
    <th>Improvement, %</th>
  </tr>
  <tr>
    <td rowspan="1">LTX2Pipeline</td>
    <td>38.34</td>
    <td>34.68</td>
    <td>🟢 9.5</td>
  </tr>
  <tr>
    <td rowspan="1">LTX2TwoStagePipeline</td>
    <td>36.84</td>
    <td>34.08</td>
    <td>🟢 7.5</td>
  </tr>
</table>

#### LTX-2.3

<table>
  <tr>
    <th>Pipeline</th>
    <th>Latency Before, ms</th>
    <th>Latency After, ms</th>
    <th>Improvement, %</th>
  </tr>
  <tr>
    <td rowspan="1">LTX2Pipeline</td>
    <td>56.93</td>
    <td>50.91</td>
    <td>🟢 10.6</td>
  </tr>
  <tr>
    <td rowspan="1">LTX2TwoStagePipeline</td>
    <td>62.35</td>
    <td>56.02</td>
    <td>🟢 10.6</td>
  </tr>
  <tr>
    <td rowspan="1">LTX2TwoStageHQPipeline</td>
    <td>❌</td>
    <td>231.26</td>
    <td>⚪ N/A</td>
  </tr>
</table>

**T2V:** latency improved by **7.6–10.9%** for the supported LTX-2 and LTX-2.3 pipelines.
**I2V:** latency improved by **7.5–10.6%**, with the largest gain observed for LTX2TwoStagePipeline on LTX-2.3.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #34218185792](https://github.com/sgl-project/sglang/actions/runs/34218185792)
Latest PR Test (Extra): :x: [Run #34218185257](https://github.com/sgl-project/sglang/actions/runs/34218185257)
Latest PR Test (AMD ROCm 7.2): :x: [Run #34218185744](https://github.com/sgl-project/sglang/actions/runs/34218185744)

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/models/dits/ltx_2.py [hunk 1]
```diff
@@ -549,7 +549,9 @@ def forward(
         num_rope_elems = num_pos_dims * 2
         # LTX-2.3 HQ is sensitive to RoPE rounding; keep frequency generation on
         # the target device instead of caching a CPU/NumPy tensor.
-        freqs_dtype = torch.float64 if self.double_precision else torch.float32
+        freqs_dtype = (
+            torch.float64 if self.double_precision and not _is_npu else torch.float32
+        )
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/models/dits/ltx_2.py]:
```suggestion
        freqs_dtype = torch.float64 if self.double_precision else torch.float32
        if _is_npu:
            logger.warning_once(f"Ascend backend doesn't support float64 now, so we convert it to float32.")
            freqs_dtype = torch.float32
```

File: python/sglang/multimodal_gen/runtime/models/dits/ltx_2.py [hunk 2]
```diff
@@ -1766,6 +1768,11 @@ def __init__(
                 hf_config.get("rope_double_precision", arch.double_precision_rope)
             )
         )
+        if rope_double_precision and _is_npu:
+            logger.warning(
+                f"Double precision is not supported by NPU. Use float32 for RoPE."
+            )
+
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/models/dits/ltx_2.py]:
```suggestion
        if rope_double_precision and _is_npu:
            logger.warning(
                f"Ascend backend does not support float64 for RoPE until now. Falling back to float32."
            )
            rope_double_precision = False

```

File: python/sglang/multimodal_gen/runtime/models/dits/ltx_2.py (hunk 1 above)
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/models/dits/ltx_2.py]:
```suggestion
        freqs_dtype = torch.float64 if self.double_precision else torch.float32
```

File: python/sglang/kernels/ops/diffusion/triton/ltx2_ada_values.py [hunk 3]
```diff
@@ -141,14 +141,14 @@ def ltx2_ada_values9(
 ) -> tuple[torch.Tensor, ...]:
     if timestep.ndim != 3:
         raise ValueError("timestep must have shape [B, S, 9 * D]")
-    if not timestep.is_cuda or timestep.dtype != torch.bfloat16:
+    if not (timestep.is_cuda or timestep.is_npu) or timestep.dtype != torch.bfloat16:
```
Comment (Makcum888e) [inline python/sglang/kernels/ops/diffusion/triton/ltx2_ada_values.py]: could we add new `current_platform.tensor_on_device(x)`

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/ltx_2/denoising.py [hunk 4]
```diff
@@ -563,7 +568,7 @@ def _ltx2_res2s_new_noise(
         noise = torch.randn(
             reference_tensor.shape,
             generator=generator,
-            dtype=torch.float64,
+            dtype=torch.float32 if _is_npu else torch.float64,
```
Comment (Makcum888e) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/ltx_2/denoising.py]: could we add new `current_platform.support_fp64()`?

Comment (e-martirosian) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/ltx_2/denoising.py]: Good idea. Set existed _float64_supported to False and reused it.

File: python/sglang/kernels/ops/diffusion/triton/ltx2_ada_values.py (hunk 3 above)
Comment (e-martirosian) [reply python/sglang/kernels/ops/diffusion/triton/ltx2_ada_values.py]: Added current_platform.tensor_on_device(x)

File: python/sglang/multimodal_gen/runtime/platforms/npu.py [hunk 5]
```diff
@@ -40,6 +41,11 @@ class NPUPlatformBase(Platform):
     dispatch_key: str = "NPU"
     device_control_env_var: str = "ASCEND_RT_VISIBLE_DEVICES"
 
+    @classmethod
+    @lru_cache(maxsize=1)
+    def is_float64_supported(cls) -> bool:
+        return False
+
```
Comment (Makcum888e) [inline python/sglang/multimodal_gen/runtime/platforms/npu.py:48]:
```suggestion

    def tensor_on_device(self, t: torch.Tensor) -> bool:
        return t.is_npu

```

File: python/sglang/multimodal_gen/runtime/platforms/interface.py [hunk 6]
```diff
@@ -426,6 +426,10 @@ def get_attn_backend(self, *args, **kwargs) -> AttentionImpl:
         attention_cls_str = self.get_attn_backend_cls_str(*args, **kwargs)
         return resolve_obj_by_qualname(attention_cls_str)
 
+    def tensor_on_device(self, t: torch.nn.Modulee) -> bool:
+        """Check if a tensor is on the current platform's device."""
+        return t.is_npu or t.is_cuda
```
Comment (Makcum888e) [inline python/sglang/multimodal_gen/runtime/platforms/interface.py]:
```suggestion
    def tensor_on_device(self, t: torch.Tensor) -> bool:
        """Check if a tensor is on the current platform's device."""
        return t.is_cuda
```

**Review by Makcum888e — APPROVED** [review]
LGTM

**Review by ping1jing2 — APPROVED** [review]
